### PR TITLE
MAINT: sparse.linalg: Correct the criterion and iteration number to show in `tfqmr.py`

### PR DESCRIPTION
**Pull request by zhaog6**
…w in `tfqmr.edp`

#### What does this implement/fix?

Corrected the convergence criterion (`tau * np.sqrt(even+1) < atol` is a better choice) and the iteration number to print to screen. The latter should be half of the number of the `for ` loops.

### CODE REVIEW

**Comment by zhaog6** [comment]
Dear @rgommers , the convergence criterion is `tau * np.sqrt(iter+1) < atol` in the paper cited, but from practical implementation, using `tau * np.sqrt(even+1) < atol` (that is PETSc's handling) will be more proper because I applied it to solve Poisson equations and convection-diffusion equations and found that it will take fewer iteration steps to reach same tolerence.

**Comment by rgommers** [comment]
Thanks for the details @zhaog6. I looked at this a bit more closely, and neither criteria makes a lot of sense to me. For the original `tau * np.sqrt(iter+1) < atol`: `atol` is constant in the `for iter in range(maxiter)` loop; `iter` increases, so `tau` decreases and it somehow needs to decrease faster than `1/sqrt(iter)` in order to ever reach convergence. These lines are how `tau` is changing:
```python
theta = np.linalg.norm(w) / tau
c = np.sqrt(1. / (1 + theta**2))
tau *= theta * c
```
Which is approximately constant for large `tau`, and ~`1/iter` for small `tau`.

With your proposed change, `np.sqrt(iter+1)` no longer increases continuously but just stays constant, jumping back and forth between `1` and `sqrt(2)`. At that point why the jumping, why not just `tau < atol`?



> (that is PETSc's handling) will be more proper because I applied it to solve Poisson equations and convection-diffusion equations and found that it will take fewer iteration steps to reach same tolerence.

I think that that's a reasonable argument: it works on several applications, and it's what PETSc does. Could you add a link to the PETSc source code or PR that's most relevant?

**Comment by zhaog6** [comment]
> For the original `tau * np.sqrt(iter+1) < atol`: `atol` is constant in the `for iter in range(maxiter)` loop; `iter` increases, so `tau` decreases and it somehow needs to decrease faster than `1/sqrt(iter)` in order to ever reach convergence.

Yes, `tau` decreases with iterations at `||w|| / sqrt(1 + ||w||^2)`.
> With your proposed change, np.sqrt(iter+1) no longer increases continuously but just stays constant, jumping back and forth between 1 and sqrt(2). At that point why the jumping, why not just tau < atol?

I think it may be a bug in PETSc, and I'll contact the author for discussion. The original convergence criterion is no problem because it is mathematically derived, that is "||r|| <= sqrrt(iter + 1) * tau < tol". Keep the original for the time being, thanks. In additional, the iteration step to print should be indeed `np.int64((iter + 1 + even) / 2)`.
> Could you add a link to the PETSc source code or PR that's most relevant?

Okay, I added the link to the source code to "Additional Information".

File: scipy/sparse/linalg/isolve/tfqmr.py [hunk 1]
```diff
@@ -145,7 +145,7 @@ def tfqmr(A, b, x0=None, tol=1e-5, maxiter=None, M=None,
         d = u + (theta**2 / alpha) * eta * d  # [1]-(5.5)
         # [1]-(5.2)
         theta = np.linalg.norm(w) / tau
-        c = np.sqrt(1. / (1 + theta**2))
+        c = 1. / np.sqrt(1 + theta**2)
```
Comment (ilayn) [inline scipy/sparse/linalg/isolve/tfqmr.py:148]: if we are touching this, better use `np.hypot(1, theta)`

File: scipy/sparse/linalg/isolve/tfqmr.py [hunk 2]
```diff
@@ -155,11 +155,11 @@ def tfqmr(A, b, x0=None, tol=1e-5, maxiter=None, M=None,
...
             callback(x)
 
-        # Convergence criteron
-        if tau * np.sqrt(iter+1) < atol:
+        # Convergence criterion
+        if tau * np.sqrt(even+1) < atol:
             if (show):
                 print("TFQMR: Linear solve converged due to reach TOL "
-                      "iterations {}".format(iter+1))
+                      "iterations {}".format(np.int64((iter+1+even)/2)))
```
Comment (ilayn) [inline scipy/sparse/linalg/isolve/tfqmr.py:162]:
```suggestion
                      f"iterations {iter+1+even)//2}"
```

**Comment by zhaog6** [comment]
Thanks @ilayn for your review. Several places have been modified.

**Comment by zhaog6** [comment]
Dear @rgommers @ilayn 
This seemingly good convergence criterion is indeed a bug, I had submitted a bug fix to PETSc team and the commit has been merged into the `release` branch. The link address: https://gitlab.com/petsc/petsc/-/merge_requests/4290
I resubmitted the PR and squeeze its commit history to facilitate to review again. Thanks.

**Comment by rgommers** [comment]
Thanks @zhaog6, this looks better to me.

> because I applied it to solve Poisson equations and convection-diffusion equations and found that it will take fewer iteration steps to reach same tolerence.

Is this still true for you?

**Comment by zhaog6** [comment]
> Is this still true for you?

Thanks for your review @rgommers . I used the incorrect convergence criterion to solve Poisson and convection-diffusion systems with Dirichlet boundary condition, and  its result is always right. This may be due to the weak nonlinearity and singularity of the equations. In fact, `tau < atol` (or `sqrt(m+1) * tau < atol (m = 0, 1)` ) does not guarantee that `norm(residual) < atol`. Although it seems to do well in the above two typical examples, we don't konw what kind of equation will go wrong because there is no guarantee of mathematical theory. However, it's always good to use `sqrt(iter+2.) * tau < atol` thank to mathematical theory in Freund's paper. So it should be use all the time.

**Comment by zhaog6** [comment]
Dear @rgommers @ilayn , If the PR is no problem, Could you merge it for me? Thanks.

**Review by rgommers — COMMENTED** [review]
Thanks for the update @zhaog6. One more question on iteration count.

**Comment by zhaog6** [comment]
Dear Ilhan @ilayn , Could you help me to review the PR when you have time? if there is any problem with this PR, please let me know. Thanks again.

**Comment by zhaog6** [comment]
Dear Ralf @rgommers 

For TFQMR algorithm, I made a modification which will be easier to understand. The actual iteration steps is a half of for-loops due to odd- and even-loop.

**Pull request closed without merging**